Added check for torch.where on CPU that both arguments have same dtype

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30662 Added check for torch.where on CPU that both arguments have same dtype**
* #30661 Make zeros argument of torch.where same dtype as other argument

Cherry picked from: https://github.com/pytorch/pytorch/pull/29081

Co-authored-by: ifedan

Differential Revision: [D18782295](https://our.internmc.facebook.com/intern/diff/D18782295)

nairbv committed:
Added check for torch.where on CPU that both arguments have same dtype

Cherry picked from: https://github.com/pytorch/pytorch/pull/29081

Co-authored-by: ifedan

[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -763,6 +763,45 @@ def test_where_bool_tensor(self):
             self.assertEqual(1, len(res))
 
+    def test_where_tensor(self):
+        def rand_tensor(size, dtype, device):
```
Could we please deduplicate the rand_tensor calls? This is the third place we have logic to create a random tensor; this looks pretty similar to the one in test_type_promotion.

nairbv replied:
as in the other review, there's plenty we could do, but I want master to match 1.3.1 first.

nairbv replied:
this is just a cherry pick from 1.3.1 to master, also need to merge to 1.4

zou3519 replied:
I'm pretty uncomfortable with having the same code duplicated. Could you please send a followup PR to deduplicate this and https://github.com/pytorch/pytorch/blob/8b29701ae505a17595a85cac05109ddf3f82e68e/test/test_type_promotion.py#L174-L187 ?

nairbv replied:
there's an issue filed. I added a reference to this in the issue tracker:
https://github.com/pytorch/pytorch/issues/30611

nairbv committed:
Update on "Added check for torch.where on CPU that both arguments have same dtype"


Cherry picked from: https://github.com/pytorch/pytorch/pull/29081

Co-authored-by: ifedan

Differential Revision: [D18782295](https://our.internmc.facebook.com/intern/diff/D18782295)

[ghstack-poisoned]
